src/herring/model.hpp

```diff
@@ -0,0 +1,255 @@
+#pragma once
+
+#include<cstddef>
+#include<functional>
+#include<iostream>
+#include<new>
+#include<numeric>
+#include<type_traits>
+#include<vector>
+
+#include<herring/omp_helpers.hpp>
+#include<herring/output_ops.hpp>
+#include<herring/tree.hpp>
+#include<herring/type_helpers.hpp>
+
+
+
+namespace herring {
+
+#ifdef __cpp_lib_hardware_interference_size
+  using std::hardware_constructive_interference_size;
+#else
+  auto constexpr hardware_constructive_interference_size = std::size_t{64};
+#endif
+
+  template <typename value_t, typename feature_index_t, typename offset_t, typename output_index_t, typename output_t>
+  struct simple_model {
+    using simple_tree_t = simple_tree<value_t, feature_index_t, offset_t, output_index_t, output_t>;
+    using lookup_tree_t = lookup_tree<value_t, feature_index_t, offset_t, output_index_t, output_t>;
+    using tree_type = std::conditional_t<
+      std::is_same_v<value_t, output_t> || std::is_same_v<output_index_t, output_t>,
+      simple_tree_t,
+      lookup_tree_t
+    >;
+    using sum_elem_type = typename is_container_specialization<output_t, std::vector>::value_type;
+
+    std::vector<tree_type> trees;
+    std::size_t num_class;
+    std::size_t num_feature;
+    element_op element_postproc;
+    row_op row_postproc;
+    float average_factor;
+    float bias;
+    float postproc_constant;
+    std::vector<bool> mutable row_has_missing;
+    bool use_inclusive_threshold;
+
+    auto precompute_missing(float const* input, std::size_t num_row) const {
+      auto result = false;
+      if (num_row > row_has_missing.size()) {
+        row_has_missing.resize(num_row);
+      }
+      for (auto row_index = std::size_t{}; row_index < num_row; ++row_index) {
+        row_has_missing[row_index] = std::transform_reduce(
+          input + row_index * num_feature,
+          input + (row_index + 1) * num_feature,
+          false,
+          std::logical_or<>(),
+          [](auto val) { return std::isnan(val); }
+        );
+        result = result || row_has_missing[row_index];
+      }
+      return result;
+    }
+
+    void predict(float const* input, std::size_t num_row, float* output, thread_count<int> nthread) const {
+      if (!precompute_missing(input, num_row)) {
+        if (!use_inclusive_threshold) {
+          predict_<false, false>(input, output, num_row, nthread);
+        } else {
+          predict_<false, true>(input, output, num_row, nthread);
+        }
+      } else {
+        if (!use_inclusive_threshold) {
+          predict_<true, false>(input, output, num_row, nthread);
+        } else {
+          predict_<true, true>(input, output, num_row, nthread);
+        }
+      }
+    }
+
+    void apply_postprocessing(
+        std::vector<sum_elem_type> const& grove_sum,
+        float* output,
+        std::size_t num_row,
+        std::size_t num_grove,
+        thread_count<int> nthread) const {
+
+      // TODO(wphicks): Precompute this
```

Comment:
Github issue?

Reply:
Decided to just resolve it after reflecting on Rory's comment. Will push up the change shortly